Handled `AnnotatedRouteNoHostException` causing the `KafkaRebalance` to get into `NotReady` state when CC is enabled form Disabled state

ShubhamRwt opened the pull request:
Signed-off-by: ShubhamRwt <srawat@redhat.com>

### Type of change

- Bugfix

### Description

This PR fixes the issue for `KafkaRebalance` going into `NotReady` state when CC is enabled from Disabled state due to the `AnnotatedRouteNoHostException`

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md
- [ ] Supply screenshots for visual changes, such as Grafana dashboards

ppatierno commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:
```diff
@@ -1362,7 +1362,9 @@ private boolean hasRebalanceAnnotation(KafkaRebalance kafkaRebalance) {
-        if (rebalanceAnnotation == KafkaRebalanceAnnotation.refresh || kafkaRebalance.getStatus().getConditions().stream().anyMatch(condition -> "ConnectException".equals(condition.getReason()))) {
+        if (rebalanceAnnotation == KafkaRebalanceAnnotation.refresh
+                || kafkaRebalance.getStatus().getConditions().stream().anyMatch(condition -> "ConnectException".equals(condition.getReason()))
+                || kafkaRebalance.getStatus().getConditions().stream().anyMatch(condition -> "AnnotatedNoRouteToHostException".equals(condition.getReason()))) {
```
Tbh I suggested to use the base class `NoRouteToHostException` as we catch the `ConnectException` and not the derived `AnnotatedConnectException`. They are part of the Java base and not Netty. @ShubhamRwt @im-konge can you try with using `NoRouteToHostException` if it works for you?

ShubhamRwt commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/cruisecontrol/CruiseControlApiImpl.java:
```diff
@@ -435,6 +436,10 @@ private void httpExceptionHandler(Promise<? extends CruiseControlResponse> resul
             // so we catch and raise a TimeoutException instead
             result.fail(new TimeoutException(t.getMessage()));
+        } else if (t instanceof NoRouteToHostException) {
+            // Netty throws a AnnotatedNoRouteToHostException when the request times out
```
Fixed @ppatierno. Sorry for not pushing it early

ppatierno reviewed (approved):
LGTM

scholzj commented:
/azp run regression

scholzj commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:
```diff
@@ -1362,7 +1362,9 @@ private boolean hasRebalanceAnnotation(KafkaRebalance kafkaRebalance) {
         // check if refresh annotation applied or request are made when CC is not active
-        if (rebalanceAnnotation == KafkaRebalanceAnnotation.refresh || kafkaRebalance.getStatus().getConditions().stream().anyMatch(condition -> "ConnectException".equals(condition.getReason()))) {
+        if (rebalanceAnnotation == KafkaRebalanceAnnotation.refresh
+                || kafkaRebalance.getStatus().getConditions().stream().anyMatch(condition -> "ConnectException".equals(condition.getReason()))
+                || kafkaRebalance.getStatus().getConditions().stream().anyMatch(condition -> "NoRouteToHostException".equals(condition.getReason()))) {
```
It is probably worth adding some comment explaining this. Why does it look for this in the conditions? Also, why does it not use the actual exception classes?

scholzj replied:
As commented before, this might deserve an explanation why this checks for the errors in conditions, because that seems wierd.

tomncooper replied:
I agree a better description is needed an that this checking could be made more robust. 

@scholzj what would be a better way to check for these types of exceptions?

scholzj replied:
@tomncooper Well, I do not know why it does what it does -> so it is hard to say what is a _better_ to do it.

But from my PoV ... the controller / operator pattern is based on retrying things in every reconciliation. So why does it need to look into the conditions what the error is? Shouldn't it just know what the error is because it just got the exception from some operation?

scholzj replied:
As I said ... if this is the best way how to deal with something, fine, but it seems unusual at best, so I think it should be explained in the comment. And I also think that this `why does it not use the actual exception classes` is a valid comment even if the logic stays.

scholzj commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/cruisecontrol/CruiseControlApiImpl.java:
```diff
@@ -435,6 +436,10 @@ private void httpExceptionHandler(Promise<? extends CruiseControlResponse> resul
             // so we catch and raise a TimeoutException instead
             result.fail(new TimeoutException(t.getMessage()));
+        } else if (t instanceof NoRouteToHostException) {
+            // Netty throws a AnnotatedNoRouteToHostException when it cannot resolve the host
```
You talk about `AnnotatedNoRouteToHostException` buit you do not seem to use it. So ids the code wrong? Or the comment wrong?

ShubhamRwt replied:
Hi Jakub, Actually we do get a `AnnotatedNoRouteToHostException` but we catch `NoRouteToHostException` instead since `AnnotatedNoRouteToHostException` (inherits from NoRouteToHostException)

scholzj replied:
So the comment is wrong?

scholzj reviewed (commented):
Should there be some unit test for this?

ShubhamRwt commented:
Added the tests @scholzj . I refactored the `createKafkaRebalance` method in `KafkaRebalanceStateMachineTest` class so that I can add Exception to the status and see the state change

scholzj commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/cruisecontrol/CruiseControlApiImpl.java:
```diff
@@ -435,6 +436,10 @@ private void httpExceptionHandler(Promise<? extends CruiseControlResponse> resul
             // so we catch and raise a TimeoutException instead
             result.fail(new TimeoutException(t.getMessage()));
+        } else if (t instanceof NoRouteToHostException) {
+            // Netty throws a AnnotatedNoRouteToHostException (inherits from NoRouteToHostException) when it cannot resolve the host
+            //so we catch and raise a NoRouteToHostException instead
```
I still don't get from this why are you not checking for `t instanceof AnnotatedNoRouteToHostException` => that is what the comment needs to explain. Because you say you do something else than what you actually do.

ShubhamRwt replied:
We are catching `NoRouteToHostException` just like we catch the ConnectException and not the derived AnnotatedConnectException. I guess in the comments I wrote the same reason that it catch and raise a `NoRouteToHostException`  instead.

scholzj replied:
But the comment does not really explain it. This seems weird:
* Why do you need to raise another exception when it inherits?
* If you need to raise the `NoRouteToHostException` instead, why do you check for `instanceof NoRouteToHostException` and not `instanceof AnnotatedNoRouteToHostException`

None of that is explained there. The comment just says what it does, but it needs to say why it does it.

ShubhamRwt replied:
Actually if you see the other if conditions there also we are doing the same thing. We try to catch the `Parent Exception`. We are doing this because `ConnectException`, `NoRouteToHostException` all these exceptions are basically part of of Java base.

scholzj replied:
I'm not sure that is a justification - that some other place does the same for sure not. And if an exception from some undesired transitive dependency is raised, it is probably a bug somewhere. But in any case, it is not what the comment says. So even if the code is kept as it is, the comment should be updated.

ShubhamRwt replied:
Yeah I agree with you. I guess its better to catch `AnnotatedNoRouteHostException` itself and then raise a `NoRouteToHostException`. I was just trying to be consistent. I will fix the comment here

scholzj replied:
I think you should find out who wrote the original code and have them explain why they did what they did.

ppatierno replied:
@tomncooper do you remember why we went this way? Was it just for handling the Java base exceptions up the chain instead of handling the Netty related ones?

tomncooper replied:
I do not remember why this was done? Git blame shows I wrote and Paolo was the last to change it. It doesn't seem to make sense to rethrow the parent Exception?

ppatierno replied:
Maybe it was done for making the code which catches that exception more general and not relying on some specific exception coming from Netty but the base one coming from Java?

tomncooper replied:
Well if you want to handle this you could still catch the more generic exception? I tried deleting this code and running the tests and nothing bad occured (except one test that exects the connection exceception not the netty one to be thrown). Maybe the system tests would show more, but this does seem to be a pointless operation?

ShubhamRwt replied:
Should I catch the exception which are thrown itself and not the parent Exceptions then @ppatierno @tomncooper . Also I will test if it works fine and if it does then maybe we can keep it that way?

tomncooper replied:
+1 for not doing this re-throwing of the parent. I tested the change and nothing broke so I am for removing this.

The only reason for doing this I can see is not exposing the dependency exception types in the API but they inherit from base Java exceptions so this is nor eeally an issue as far as I am concerned.

scholzj commented:
@ShubhamRwt Can you please fix the build and the DCO signoff?

ShubhamRwt commented:
Sure. Actually we need some discussion on this PR and yesterday Tom and I decided to have a meeting today regarding this

ShubhamRwt commented:
I have made the new changes @ppatierno @tomncooper

ppatierno commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/cruisecontrol/CruiseControlApiImpl.java:
```diff
@@ -435,11 +436,12 @@ private void httpExceptionHandler(Promise<? extends CruiseControlResponse> resul
-            // so we catch and raise a ConnectException instead
-            result.fail(new ConnectException(t.getMessage()));
-        } else {
+            // so we catch and raise a CruiseControlRestException instead since
```
raise a CruiseControlRetriableConnectException not CruiseControlRestException

ppatierno commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceStateMachineTest.java:
```diff
@@ -96,12 +97,13 @@ public void resetServer() {
                                                 String currentStatusSessionID,
                                                 String userAnnotation,
-                                                KafkaRebalanceSpec rebalanceSpec) {
+                                                KafkaRebalanceSpec rebalanceSpec, String exceptionString) {
```
I see it sets the reason on the condition, why not calling it `reason` instead of `exceptionString`?

ppatierno commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceStateMachineTest.java:
```diff
@@ -97,11 +98,12 @@ public void resetServer() {
      * @param userAnnotation An annotation to be applied after the reconcile has started, for example "approve" or "stop".
      * @param rebalanceSpec A custom rebalance specification. If null a blank spec will be used.
      * @return A KafkaRebalance instance configured with the supplied parameters.
+     * @return Reason for the condition
```
I think it's wrong. You have got two `@return` I guess you meant to add a `@param reason`

ShubhamRwt replied:
Oops Sorry, Let me correct it

ShubhamRwt replied:
Fixed

ppatierno reviewed (approved):
Just a final nit, but LGTM overall to me.

tomncooper commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/cruisecontrol/CruiseControlRetriableConnectException.java:
```diff
@@ -0,0 +1,12 @@
+ */
+package io.strimzi.operator.cluster.operator.resource.cruisecontrol;
+
+public class CruiseControlRetriableConnectException extends RuntimeException {
```
I wonder if a better name for this would have been `CruiseControlRetriableConnectionException` to be explicit and avoid any possible confusion with Kafka Connect.

ppatierno replied:
Don't know ... I kind of like the relationship with the `ConnectException` we catch in the code to make this.

scholzj replied:
I would agree with @tomncooper here. The `CruiseControlRetriableConnectException` seems weird to me as well. Double confusing with Kafka Connect being around.

scholzj replied:
Also, it is thrown for `NoRouteToHostException` as well, not only for `ConnectException` which might add to another confusion.

tomncooper commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/cruisecontrol/CruiseControlApiImpl.java:
```diff
@@ -435,11 +436,12 @@ private void httpExceptionHandler(Promise<? extends CruiseControlResponse> resul
-            result.fail(new ConnectException(t.getMessage()));
-        } else {
+            // so we catch and raise a CruiseControlRetriableConnectException instead since
+            result.fail(new CruiseControlRetriableConnectException(t.getMessage()));
```
So I think this is the right approach. However, now the KRAO is expecting **any** implementation of the `CruiseControlApi` interface to fail a future, if it is a transient connection error, with a `CruiseControlRetriableConnectException` from many/all its methods. 

Ideally, we would put this exception handler in an an abstract class so other implementation could use it, but maybe it would be acceptable to simply update the documentation of the interface class to explicitly say that any transient connection error should be failed with a `CruiseControlRetriableConnectException`?

@ppatierno WDYT?

ppatierno replied:
>  other implementation could use it

Which "other implementation" are you referring to?

scholzj commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/cruisecontrol/CruiseControlApiImpl.java:
```diff
@@ -435,11 +436,12 @@ private void httpExceptionHandler(Promise<? extends CruiseControlResponse> resul
-            result.fail(new ConnectException(t.getMessage()));
-        } else {
+            // so we catch and raise a CruiseControlRetriableConnectException instead since
+            result.fail(new CruiseControlRetriableConnectException(t.getMessage()));
```
Should this inherit the exception instead of copying only the message? That way, it should keep the whole stack etc. and I do not see what problems would it cause.

```suggestion
            result.fail(new CruiseControlRetriableConnectException(t));
```

ShubhamRwt replied:
I tried doing this but I was getting some issues since the message which is being set in the condition seems to come out as `null` if I do that so probably up in the stack the message we are passing is being used

scholzj replied:
But reusing the exception is important since that gives you the stack trace etc. I do not understand how the message being null relates to it.

ShubhamRwt commented:
@tomncooper @ppatierno Does this PR need more changes?

ppatierno commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:
```diff
@@ -1362,7 +1362,10 @@ private boolean hasRebalanceAnnotation(KafkaRebalance kafkaRebalance) {
+        // We check for `CruiseControlRetriableConnectException` if request are made when CC was not active since not handling this case will move the KafkaRebalance to `NotReady`
+        // and it will stay in `NotReady` state until we apply `refresh`
+        if (rebalanceAnnotation == KafkaRebalanceAnnotation.refresh
+                || kafkaRebalance.getStatus().getConditions().stream().anyMatch(condition -> "CruiseControlRetriableConnectException".equals(condition.getReason()))) {
```
should it be "CruiseControlRetriableConnectionException" now? Even in the comment above?

ppatierno commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceStateMachineTest.java:
```diff
@@ -60,6 +60,7 @@ public class KafkaRebalanceStateMachineTest {
     private static final String RESOURCE_NAME = "my-rebalance";
     private static final String CLUSTER_NAMESPACE = "cruise-control-namespace";
     private static final String CLUSTER_NAME = "kafka-cruise-control-test-cluster";
+    private static final String REASON = "CruiseControlRetriableConnectException";
```
Using "CruiseControlRetriableConnectionException" as reason here as well?

ppatierno commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceStateMachineTest.java:
```diff
@@ -873,6 +877,43 @@ private void testStateWithWrongAnnotation(Vertx vertx, VertxTestContext context,
+    }
+
+
+    private void testStateWithException(Vertx vertx, VertxTestContext context,
```
this methos seems to be exactly the same as `testStateWithWrongAnnotation` but just with a different name. How is it testing that the connection exception is recovered?

ShubhamRwt replied:
The `testStateWithException` method is always taking the current state as `ProposalReady` and checking the transition from `NotReady` to `ProposalReady` while for the `testStatewithWrongAnnotations` we have to check for multiple next states since its being used by multiple other tests

ShubhamRwt replied:
@ppatierno I have removed the `test` which I was adding named `testStateWithNoRouteToHostEception` since now `CruiseControlRetriableConnectionException` is being thrown in both cases(ConnectExcpetion && testStateWithNoRouteToHostEception) so I guess only one test is enough

ppatierno commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceStateMachineTest.java:
```diff
@@ -873,6 +877,43 @@ private void testStateWithWrongAnnotation(Vertx vertx, VertxTestContext context,
+    }
+
+    @Test
+    public void testReadyAfterConnectException(Vertx vertx, VertxTestContext context) throws IOException, URISyntaxException {
```
Let's use "ConnectionException" in the method name for consistency

ShubhamRwt replied:
Should we name it 'testReadywithCruiseControlRetriableException' itself? Since we are addressing two errors under the same exception?

ppatierno reviewed (approved):
LGTM overall.

scholzj commented:
/azp run regression
